core/client/hammer/hammer.go

```diff
@@ -76,125 +93,159 @@ func New(ctx context.Context, dial DialFunc, callOptions CallOptions,
+	requests = genRequests(ctx, c.HistoryQPS, c.HistoryBatchSize, c.HistoryCount, c.Duration)
+	handlers = make([]ReqHandler, 0, len(workers))
+	for _, w := range workers {
+		handlers = append(handlers, w.historyOp)
+	}
+	generateReport(ctx, requests, handlers)
 
-	return outputs
+	return nil
 }
 
-// gen produces a channel of integers from 0 up to count-1 and then closes.
-func gen(ctx context.Context, count int) <-chan int {
-	jobs := make(chan int)
+func genRequests(ctx context.Context, qps, batch, count int, duration time.Duration) <-chan request {
+	inflightReqs := make(chan request)
 	go func() {
-		defer close(jobs)
+		cctx, cancel := context.WithTimeout(ctx, duration)
+		defer cancel()
+		defer close(inflightReqs)
+
+		rateLimiter := rate.NewLimiter(rate.Limit(qps), qps)
 		for i := 0; i < count; i++ {
-			select {
-			case <-ctx.Done():
+			userIDs := make([]string, 0, batch)
+			for j := 0; j < batch; j++ {
+				userIDs = append(userIDs, fmt.Sprintf("user_%v", i*batch+j))
+			}
+			if err := rateLimiter.Wait(cctx); err != nil {
 				return
-			case jobs <- i:
+			}
+			inflightReqs <- request{
+				UserIDs:   userIDs,
+				BatchSize: batch,
 			}
 		}
 	}()
-	return jobs
+	return inflightReqs
 }
 
-// printStats prints stats repeatedly every period until outputs closes.
-func (h *Hammer) printStats(period time.Duration, outputs <-chan error) {
-	startTime := time.Now()
-	var totalRequests float64
-
-	// Print stats asyncronously until refresh closes.
-	refresh := time.NewTicker(period)
-	stop := make(chan struct{})
-	var wg sync.WaitGroup
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		for {
-			select {
-			case <-refresh.C:
-				rate := totalRequests / time.Since(startTime).Seconds()
-				log.Printf("QPS: %v, \tRequests: %v, \tWorkers: %v", rate, totalRequests, h.workers)
-			case <-stop:
-				rate := totalRequests / time.Since(startTime).Seconds()
-				log.Printf("QPS: %v, \tRequests: %v, \tWorkers: %v", rate, totalRequests, h.workers)
-				refresh.Stop()
-				return
-			}
-		}
-	}()
+type worker struct {
+	*Hammer
+	client *client.Client
+}
 
-	// Collect outputs until output closes.
-	for err := range outputs {
-		totalRequests++
+func (h *Hammer) newWorkers(n int) ([]worker, error) {
+	workers := make([]worker, n)
+	for i := range workers {
+		// Give each worker it's own client.
+		client, err := client.NewFromConfig(h.ktCli, h.domain)
 		if err != nil {
-			log.Printf("writeOp(): %v", err)
+			return nil, err
 		}
+
+		workers[i] = worker{Hammer: h, client: client}
 	}
-	close(stop)
-	wg.Wait()
+	return workers, nil
 }
 
-func (h *Hammer) worker(ctx context.Context, jobs <-chan int, outputs chan<- error) {
-	client, err := client.NewFromConfig(h.ktCli, h.config)
-	if err != nil {
-		glog.Errorf("NewFromConfig(): %v", err)
-		return
+// batchWriteOp queues many user mutations, waits, and then verifies them all.
+func (w *worker) writeOp(ctx context.Context, req *request) error {
+	users := make([]*tpb.User, 0, req.BatchSize)
```

Comment:
`len(req.UserIDs)` not `req.BatchSize` ?

Reply:
Yes, thank you.